anker/src/state.rs

```diff
@@ -90,6 +90,74 @@ pub struct HistoricalStSolPrice {
     pub st_sol_price_in_ust: MicroUst,
 }
 
+#[repr(C)]
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Default,
+    BorshDeserialize,
+    BorshSerialize,
+    BorshSchema,
+    Eq,
+    PartialEq,
+    Serialize,
+)]
+pub struct HistoricalStSolPriceArray(pub [HistoricalStSolPrice; POOL_PRICE_NUM_SAMPLES]);
+
+impl HistoricalStSolPriceArray {
+    /// Create new `HistorialStSolPriceArray` with slot 0 and 1 UST in each
+    /// position of the array.
+    pub fn new() -> Self {
+        HistoricalStSolPriceArray(
+            [HistoricalStSolPrice {
+                slot: 0,
+                st_sol_price_in_ust: MicroUst(1_000_000),
+            }; 5],
+        )
+    }
+
+    /// Get last price from the array.
+    pub fn last(&self) -> HistoricalStSolPrice {
+        self.0[POOL_PRICE_NUM_SAMPLES - 1]
+    }
+
+    /// Insert `st_sol_price_in_ust` at the end of the array and rotate it.
+    pub fn insert_and_rotate(&mut self, slot: Slot, st_sol_price_in_ust: MicroUst) {
+        // Maintain the invariant that samples are sorted by ascending slot number.
+        // The sample at index 0 is the oldest, so we remove it (well, move it to the
+        // end to be overwritten), and move everything else closer to the beginning
+        // of the array. Then we overwrite the last element with the current price
+        // and slot number, and we confirmed above that that slot number is larger
+        // than the slot number of the sample before it.
+        self.0.rotate_left(1);
+        self.0[POOL_PRICE_NUM_SAMPLES - 1].slot = slot;
+        self.0[POOL_PRICE_NUM_SAMPLES - 1].st_sol_price_in_ust = st_sol_price_in_ust;
+    }
+
+    /// Calculate the minimum amount we are willing to pay for the `StLamports`
+    /// rewards based on the median price from the historical price information.
+    pub fn calculate_minimum_price(
+        &self,
+        rewards: StLamports,
+        sell_rewards_min_out_bps: u64,
+    ) -> Result<MicroUst, ArithmeticError> {
+        // Get median historical price.
+        let median_price = self.0[POOL_PRICE_NUM_SAMPLES / 2];
```

Comment:
This is not the median price! The elements are sorted chronologically, not by ascending price!

I made a function for this, but I see I did not put it in #511. Now I remember what I was doing :sweat_smile: I wrote this program to find a function that returns the median of 5 numbers in 7 transpositions:

<details>
<summary>Program</summary>

```rust
/// Call `f` on all permutations of `candidates`.
///
/// Set `current = vec![]` at the top-level call.
/// Exits early if `f` returns `false`.
fn permutations<F: Fn(&[u32]) -> bool>(
    current: &mut Vec<u32>,
    candidates: &mut Vec<u32>,
    f: &F,
) -> bool {
    for i in 0..candidates.len() {
        let x = candidates.remove(i);
        current.push(x);
        if candidates.len() == 0 {
            let is_ok = f(&current[..]);
            if !is_ok {
                return false;
            }
        } else {
            let is_ok = permutations(current, candidates, f);
            if !is_ok {
                return false;
            }
        }
        current.pop();
        candidates.insert(i, x);
    }
    true
}

fn main() {
    for a0 in 0..5 {
    for a1 in (a0 + 1)..5 {
        for b0 in 0..5 {
        for b1 in (b0 + 1)..5 {
            for c0 in 0..5 {
            for c1 in (c0 + 1)..5 {
                for d0 in 0..5 {
                for d1 in (d0 + 1)..5 {
                    for e0 in 0..5 {
                    for e1 in (e0 + 1)..5 {
                        for f0 in 0..5 {
                        for f1 in (f0 + 1)..5 {
                            for g0 in 0..5 {
                            for g1 in (g0 + 1)..5 {
                                let is_ok = permutations(
                                    &mut vec![],
                                    &mut vec![0, 1, 2, 3, 4],
                                    &|xs| {
                                        let mut ys: Vec<u32> = xs.iter().cloned().collect();
                                        let mut cross = |i, j| {
                                            if ys[i] > ys[j] {
                                                ys.swap(i, j);
                                            }
                                        };

                                        cross(a0, a1);
                                        cross(b0, b1);
                                        cross(c0, c1);
                                        cross(d0, d1);
                                        cross(e0, e1);
                                        cross(f0, f1);
                                        cross(g0, g1);

                                        // Continue searching if this sequence of
                                        // crossings did find the median (2) of this
                                        // permutation. Otherwise abort early.
                                        ys[2] == 2
                                    },
                                );

                                if is_ok {
                                    println!("cross({}, {});", a0, a1);
                                    println!("cross({}, {});", b0, b1);
                                    println!("cross({}, {});", c0, c1);
                                    println!("cross({}, {});", d0, d1);
                                    println!("cross({}, {});", e0, e1);
                                    println!("cross({}, {});", f0, f1);
                                    println!("cross({}, {});", g0, g1);
                                    return
                                }
                            }}
                        }}
                    }}
                }}
            }}
        }}
    }}
}
```

</details>

It outputs

```rust
cross(0, 1);
cross(2, 3);
cross(0, 2);
cross(1, 3);
cross(1, 2);
cross(1, 4);
cross(2, 4);
```
where `cross(i, j)` is defined as

```rust
let mut cross = |i, j| {
    if ys[i] > ys[j] {
        ys.swap(i, j);
    }
};
```
After this, `ys[2]` contains the median. I would also add a test that generates all 5^5 possible sequences involving the numbers `0..5` and exhaustively check that the function returns the correct median by comparing it against `.sort()`.

Reply:
I was with the impression that we get the median from the array, that would be the median of the chronologically sorted elements.

Reply:
But why not just sort it 😬 ?

Reply:
> But why not just sort it

NIH-ism and premature optimization on my part :grimacing: Getting the median doesn’t require a full sort, and for small arrays of known size, sorting networks can do better than a generic sorting algorithm, and for median it is possible to do even better than a sorting network.

But yeah I probably shouldn’t have gone down that rabbit hole ... You are right, we should just call `.sort()`, if we run into the compute limit we can use a specialized median-of-5.